File: Adapter/Adapter.java
public class Adapter {
    private class USB_A {
        public void USB_A_interface() {
            System.out.println("USB-A interface");
        }
    }

    private class USB_C {
        public void USB_C_interface() {
            System.out.println("USB-C interface");
        }
    }
    
    private class USB_C_Adapter extends USB_C {
        private USB_A usb_a;

        public USB_C_Adapter(USB_A usb_a) {
            this.usb_a = usb_a;
        }
        
        @Override
        public void USB_C_interface() {
            // do something here
            System.out.print("Originally: ");
            usb_a.USB_A_interface();
            System.out.println("Now it's USB-C interface!");
        }
    }
    
    private class Macbook {
        private USB_C usb_c;
        public Macbook(USB_C usb_c) {
            this.usb_c = usb_c;
        }
        void charge() {
            usb_c.USB_C_interface();
        }
    }
    
    public void testAdapter() {
        Macbook macbook = new Macbook(new USB_C_Adapter(new USB_A()));
        macbook.charge();
    }
    

    public static void main(String[] args) {
        Adapter adapter = new Adapter();
        adapter.testAdapter();
    }
}

File: Adapter/Adapter.ts
// Adapter/USB_A.ts
class USB_A {
    USB_A_interface(): void {
        console.log("USB-A interface");
    }
}

// Adapter/USB_C.ts
class USB_C {
    USB_C_interface(): void {
        console.log("USB-C interface");
    }
}

// Adapter/USB_C_adapter.ts
class USB_C_adapter extends USB_C {
    private usb_a: USB_A;

    constructor(usb_a: USB_A) {
        super();
        this.usb_a = usb_a;
    }

    USB_C_interface(): void {
        console.log("Originally: ");
        this.usb_a.USB_A_interface();
        console.log("Now it's: ");
        console.log("USB-C interface");
    }
}

class Macbook {
    private usb_c: USB_C;

    constructor(usb_c: USB_C) {
        this.usb_c = usb_c;
    }

    charge(): void {
        this.usb_c.USB_C_interface();
    }
}

// Client
const macbook = new Macbook(new USB_C_adapter(new USB_A()));
macbook.charge();